plutus-scb/src/Cardano/Protocol/ChainEffect.hs

```diff
@@ -0,0 +1,39 @@
+{-# LANGUAGE DataKinds     #-}
+{-# LANGUAGE GADTs         #-}
+{-# LANGUAGE LambdaCase    #-}
+{-# LANGUAGE TypeOperators #-}
+
+module Cardano.Protocol.ChainEffect where
+
+import           Control.Concurrent.STM
+import           Control.Monad.Freer
+import           Control.Monad.Freer.Reader
+import           Control.Monad.Freer.State
+import           Control.Monad.Freer.Writer
+import           Control.Monad.IO.Class
+
+import           Ledger                     (Tx (..))
+import qualified Wallet.Emulator.Chain      as EC
+
+type ChainEffs = '[State EC.ChainState, Writer [EC.ChainEvent], Reader (TQueue Tx)]
+
+handleControlChain
+  :: ( Members ChainEffs effs )
+  => Eff (EC.ChainControlEffect ': effs) ~> Eff effs
+handleControlChain = interpret $ \case
+  EC.ProcessBlock -> EC.handleControlChain EC.processBlock
+
+{- | This effect interprets queuing Txs by *also* sending them
+     to a `TQueue`, communicating with the server node. -}
```

Comment:
Not necessarily. Right now you have two views of the chain. One is local (the PAB view) and one is global (the Cardano node view). We may wish to specify properties that deal with synchronisation of the two, in which case we will need both.

Right now, the local state is what we are using, and we assume that the local and global states are always in sync (this could be one of the properties we may want to check, eventually).

Reply:
But the management of our local view would be handled by the client dealing with notifications we get from the server, right? Whereas this does it when you send the tx?

Reply:
This updates the local view, saying that there is a new transaction in the tx pool, after sending it to the tx queue.

We could update the local tx pool only when the Cardano node confirms that the tx has been received (should we?), but for now my assumption is that once the tx hits the queue the transaction is in the Cardano node's tx pool.

Reply:
Well, do we actually need to maintain a local view of what's in the pool? Or is that just part of the book-keeping we do for our fake blockchain? 

I don't think we get notifications of things going into the pool, so I guess if we want to track this we'd have to do here indeed. I guess I'd just rename it then: it's no longer the pool, it's the things we know we've sent to the pool (e.g. it won't include stuff other people have sent).

Reply:
For now, I would suggest adding a comment in the handler that explains the situation, and following that have a conversation about this (add a GitHub issue to track it).

My main concern is that this data structure is used in a lot of effects and handlers. Making modifications to it may not be trivial and may pose architectural concerns. It is a distinction we make in one interpretation of the effects, and I am not sure if this should bubble up to the effects themselves (I think effects should be kept as simple and agnostic of handlers as possible).

Reply:
Created https://github.com/input-output-hk/plutus/issues/2141 to track this.